### PR TITLE
Implement NAQANet as a semantic parser

### PR DESCRIPTION
**Pull request by kl2806**
@matt-gardner  Still WIP, opening a PR so you can have a look. Thanks!

### CODE REVIEW

File: allennlp/models/semantic_parsing/drop/naqanet.py [hunk 1]
```diff
@@ -78,11 +93,31 @@ def __init__(self, vocab: Vocabulary,
...
             self._action_biases = Embedding(num_embeddings=num_actions, embedding_dim=1)
         self._action_embedder = Embedding(num_embeddings=num_actions, embedding_dim=action_embedding_dim)
 
+        self._decoder_beam_search = decoder_beam_search
+        self._max_decoding_steps = max_decoding_steps
+        
+        self._transition_function = LinkingTransitionFunction(encoder_output_dim=encoding_out_dim,
+                                                              action_embedding_dim=action_embedding_dim,
+                                                              input_attention=input_attention,
+                                                              predict_start_type_separately=False,
```
Comment (matt-gardner) [inline allennlp/models/semantic_parsing/drop/naqanet.py]: I think you should set this to `True`, and remove the START_SYMBOL.

File: allennlp/models/semantic_parsing/drop/naqanet.py [hunk 2]
```diff
@@ -152,28 +189,100 @@ def forward(self,  # type: ignore
...
-        # search, multiply the program probability by the answer probability, then sum across all
-        # programs.
-
-        # If answer is given, compute the loss.
-        # if answer_as_passage_spans is not None or answer_as_question_spans is not None \
-                # or answer_as_add_sub_expressions is not None or answer_as_counts is not None:
-            # all_log_marginal_likelihoods = torch.stack(log_marginal_likelihood_list, dim=-1)
-            # all_log_marginal_likelihoods = all_log_marginal_likelihoods + answer_ability_log_probs
-            # marginal_log_likelihood = util.logsumexp(all_log_marginal_likelihoods)
+            initial_rnn_state.append(RnnStatelet(hidden_state=final_encoder_output[i],
```
Comment (matt-gardner) [inline allennlp/models/semantic_parsing/drop/naqanet.py]: Make the initial hidden state be a `concat([passage_vector, question_vector])`.  The way you've done this here, you're not using the passage as an input to which program you should pick, whereas NAQANet does use this.

File: allennlp/models/semantic_parsing/drop/naqanet.py [hunk 3]
```diff
@@ -152,28 +189,100 @@ def forward(self,  # type: ignore
...
+                                                 memory_cell=memory_cell[i], 
+                                                 previous_action_embedding=self._first_action_embedding,
```
Comment (matt-gardner) [inline allennlp/models/semantic_parsing/drop/naqanet.py]: I'm not sure why you're not getting gradients to this parameter; from what I can tell, you should be.

File: allennlp/models/semantic_parsing/drop/naqanet.py (hunk 1 above)
Comment (kl2806) [reply allennlp/models/semantic_parsing/drop/naqanet.py]: I tried doing this but executing requires the START SYMBOL (https://github.com/allenai/allennlp/blob/master/allennlp/semparse/domain_languages/domain_language.py#L329). Should we remove this check in domain languages, or postprocess our actions here to include `[START_SYMBOL -> Answer]`?

Comment (matt-gardner) [reply allennlp/models/semantic_parsing/drop/naqanet.py]: Yeah, one option is to just prepend that action before calling `execute_action_sequence`.
